### PR TITLE
Multiple small changes

### PR DESCRIPTION
**Pull request by hybras**
- Modularize OS specific code
- Streamline cfg macro usage
- Use which to check for open programs
- use explorer on windows

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -183,7 +87,12 @@ pub fn with_in_background<T: AsRef<OsStr> + Sized>(
...
     use std::io;
-    use std::os::windows::ffi::OsStrExt;
+    use std::os::windows::{ffi::OsStrExt, process::ExitStatusExt};
+    use std::process::{Command, ExitStatus};
+    use std::ptr;
+
+    use winapi::ctypes::c_int;
+    use winapi::um::shellapi::ShellExecuteW;
 
     pub fn convert_path(path: &OsStr) -> io::Result<Vec<u16>> {
```
Comment (Byron) [inline src/lib.rs:97]: I believe this function doesn't have to be `pub` anymore.

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution - I think it looks great and is a massive improvement over what was there.

Personally I can only test MacOS and thus can't verify that `explorer` will work, but I take your word for it.
As far as I can tell everything else is a refactor, so should be alright.

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,3 @@
 target
 Cargo.lock
+.vscode
```
[the remark this answers is not in the document]
Comment (hybras) [reply .gitignore:3]: My mistake, I did this out of habit

File: src/lib.rs (hunk 1 above)
Comment (hybras) [reply src/lib.rs]: fixed in 3507b55

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (hybras) [reply .gitignore]: fixed in 021bb15

**Comment by hybras** [comment]
> Personally I can only test MacOS and thus can't verify that `explorer` will work, but I take your word for it.

My pr was intended to simply be a simple refactor. Little did I know... Now it is 10 commits long, though only 2 change functionality and some are internal bookkeeping I should not have committed. 😝

**Comment by Byron** [comment]
Fantastic, thanks so much! Version 1.5 was just released to crates.io.

I can only encourage you to keep doing this - it's probably a good learning opportunity to look at crates and refactor them in the process. At least this worked for me in the past <cough>GitPython<cough> 😅.

All the best!

**Comment by hybras** [comment]
Thx for ur feedback and advice

**Comment by wez** [comment]
Would you mind adding some context on why using explorer.exe rather than the windows shell APIs is desirable?

**Comment by hybras** [comment]
> Would you mind adding some context on why using explorer.exe rather than the windows shell APIs is desirable?

Sure. 

- It is easier to understand what's happening / more readable
- it followed the patterns set in macOS, unix of calling out to cli (tho this is more style choice)
- we removed a usage of `unsafe`
  - I admit I don't understand what invariants are required to uphold safety. It was not documented here nor in winapi.

**Comment by wez** [comment]
FWIW, the shell APIs are the defined way to open documents on windows.  I'm not sure that explorer.exe is even guaranteed to be accessible in some Windows deployments (admittedly rare).

**Comment by hybras** [comment]
Should I revert the [commit](https://github.com/Byron/open-rs/pull/26/commits/45454254b6e07fd88e398e8de86b55863f369373) and open a new pr? I'm not sure what is the best course of action. Does this crate intend to support such use cases? No problem if so

**Comment by wez** [comment]
Personally, I think it's better to use the shell API; it feels conceptually safer than running a process with an arbitrary argument.

However, I don't speak for this project, and I can always just put that same code in my downstream project, which already has a lot of windows specific functionality.

**Comment by Byron** [comment]
> Should I revert the [commit](https://github.com/Byron/open-rs/pull/26/commits/45454254b6e07fd88e398e8de86b55863f369373) and open a new pr? I'm not sure what is the best course of action. Does this crate intend to support such use cases? No problem if so

Yes, let’s do that, and a new PR is welcome. I think I will yank 1.4 and republish a patch as using explorer might break some users. It’s probably better to stick to the tried and trusted and not take any unnecessary risk.

My apologies for the lapse in judgment and thanks to @hybras for making the PR happen, and of course for the refactor earlier which is great to have :).

**Comment by Byron** [comment]
1.5.0 was yanked and 1.5.1 was released in its place, restoring the previous implementation. Thanks a lot for your help @hybras  and @wez for your suggestion.

**Comment by hybras** [comment]
@wez Where can I learn more about windows api's (it seems to be an umbrella term for a massive amount of stuff). I've been on macOS and linux my whole life, I'm still learning "win fu." Google searches reveal api docs, but not many explanations of how to do... anything

**Comment by wez** [comment]
@hybras there are certainly a great many windows specific APIs to choose from and learn about!  My knowledge of them has been accumulating since the late '90s, but I don't consider myself an expert!

I find that it is easiest to pick a topic/area that you want to play with and try reading around; Microsoft has a lot of documentation that explains concepts around sets of related win32 APIs.  The documentation is excellent in that regard, but they frequently move things around and invalidate links which can make it harder when eg: following threads from stackoverflow.

Stackoverflow is pretty decent resource for answering questions like "how do I do X in win32"; you'll usually get a selection of results that can point you in the right direction, but not always get *the* correct answer.

A really excellent resource is Raymond Chen's blog [The Old New Thing](https://devblogs.microsoft.com/oldnewthing/); there's a lot of details on the right vs. wrong way to do various things on windows, and a lot of explanations that you won't find in the official docs.

**Pull request merged**